File: 20-valid-parentheses/20-valid-parentheses.cpp
const isValid = (s) => {
    const map = {
        '(':')',
        '{':'}',
        '[':']',        
    }
    let stack = []
    for (let c of s) {
        if (c in map){ 
            stack.push(c)
        } else if (map[stack.pop()] !== c) {
            return false
        }
    }
    return !stack.length
};




// function isValid(s, r = /\(\)|\[\]|\{\}/g) {
//     while (s.match(r)) s = s.replace(r, '')
//     return !s
// }

File: 20-valid-parentheses/20-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(char c : s.toCharArray()){
            if(c=='(' || c=='[' || c=='{'){
                stack.push(c);
            }else{
                if(stack.isEmpty()) return false;
                if(c==']' && stack.peek()!='[') return false;
                if(c==')' && stack.peek()!='(') return false;
                if(c=='}' && stack.peek()!='{') return false;
                
                // all good then 
                stack.pop();
            }
        }
        
        return stack.isEmpty();
    }
}